File: cousinsTree.cpp
// Time Complexity :O(N) where n in the number of nodes
// Space Complexity : O(N)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
        int xLevel = 0;
        int yLevel = 0;
        TreeNode *xParent;
        TreeNode *yParent;
        bool xPresent = false;
        bool yPresent = false;
    
public:
    /*bool isCousins(TreeNode* root, int x, int y) {
        //DFS
        dfs(root,nullptr,0,x,y);
        return (xLevel == yLevel) && (xParent != yParent);
    }
    void dfs(TreeNode* root, TreeNode* parent, int level, int x, int y){
        //base
        if(!root) return;
        //logic
        if(root->val == x){
            xLevel = level;
            xParent = parent;
        }
        if(root-> val == y){
            yLevel = level;
            yParent = parent;
        }
        //recursion
        dfs(root->left,root,level+1,x,y);
        dfs(root->right,root,level+1,x,y);
    }*/
    bool isCousins(TreeNode* root, int x, int y){
        //BFS
        int size;
        queue<pair<TreeNode*,TreeNode*>> q;
        pair<TreeNode*,TreeNode*> curr;
        q.push({root,nullptr});
        while(!q.empty()){
            size = q.size();
            for(int i = 0;i< size;i++){
                curr = q.front();
                q.pop();
                if(curr.first->val == x){
                    xPresent = true;
                    xParent = curr.second;
                } 
                if(curr.first->val == y){
                    yPresent = true;
                    yParent = curr.second;
                }
                if(curr.first->left) q.push({curr.first->left,curr.first});
                if(curr.first->right) q.push({curr.first->right,curr.first});
            }
            if(xPresent && yPresent){
                if(xParent != yParent)
                    return true;
            }
            if(xPresent || yPresent) return false;
        }
        return false;
    }
};

File: CousinsTree.cs
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsCousins(TreeNode root, int x, int y) {        
        var q = new Queue<TreeNode>();
        var pq = new Queue<TreeNode>();
        q.Enqueue(root);
        pq.Enqueue(null);
        while(q.Count > 0)
        {
            var count = q.Count;
            bool x_level = false; TreeNode x_parent = null;
            bool y_level = false; TreeNode y_parent = null;
            for(int i = 0; i< count; i++)
            {               
                var node = q.Dequeue();
                var pNode = pq.Dequeue();
                if(node.val == x)
                {
                    x_parent = pNode;
                    x_level = true;
                }
                if(node.val == y)
                {
                    y_parent = pNode;
                    y_level = true;
                }
                if(node.left != null)
                {
                    q.Enqueue(node.left);
                    pq.Enqueue(node);
                }
                if(node.right != null)
                {
                    q.Enqueue(node.right);
                    pq.Enqueue(node);
                }
            }
            
            if( x_level && y_level)
                return x_parent != y_parent;
            
            if(x_level || y_level)
                return false;
        }  
        return false;     
    }   
}
